lib/resolve-package-path.js

```diff
@@ -0,0 +1,137 @@
+var path = require('path');
+var pathRoot = require('path-root');
+
+/*
+ * Resolve the real path for a file. Return null if does not
+ * exist or is not a file or FIFO, return the real path otherwise.
+ *
+ * Cache the result in the passed-in cache for performance, 
+ * keyed on the filePath passed in.
+ *
+ * @private
+ * @method _getRealFilePath
+ * @param cache the Cache object to cache the real path in, keyed
+ * by filePath.
+ * @param {String} filePath the (not necessarily resolved) 
+ * path to the file of interest
+ * @return {String} real path or null
+ */
+function _getRealFilePath(cache, filePath) {
+  var realPath = cache.get(filePath);
+
+  if (realPath) {
+    return realPath;
+  }
+
+  try {
+    var stat = fs.statSync(filePath);
+
+    if (stat.isFile() || stat.isFIFO()) {
+      realPath = fs.realpathSync(filePath);
+    } else {
+      realPath = null; 
+    }
+  } catch (e) {
+    if (
+      e !== null &&
+      typeof e === 'object' &&
+      (e.code === 'ENOENT' || e.code === 'ENOTDIR')
+    ) {
+      realPath = null; // NO PATH FOUND
+    } else {
+      throw e;
+    }
+  }
+
+  cache.set(filePath, realPath);
+
+  return realPath;
+}
+
+/*
+ * Given a package 'name' and starting directory, resolve to a real (existing) file path.
+ *
+ * Do it similar to how it is done in resolve.sync() - travel up the directory hierarchy,
+ * attaching 'node-modules' to each directory and seeing if the directory exists and
+ * has the relevant 'package.json' file we're searching for. It is *much* faster than
+ * resolve.sync(), because we don't test that the requested name is a directory.
+ * This routine assumes that it is only called when we don't already have
+ * the cached entry.
+ *
+ * @param realFilePathCache the cache containing the real paths corresponding to
+ * various file paths (which may or may not be already resolved). This is only used
+ * if passed through to _getRealFilePath.
+ * @param name the 'name' of the module, i.e. x in require(x), but with
+ *   '/package.json' on the end. It is NOT a directory (so we don't have to do the
+ *   directory checks that resolve.sync does).
+ *   NOTE: because this is an internal routine, for speed it does not check 
+ *   that '/package.json' is actually the end of the name.
+ * @param startPath the directory (non-empty, and valid) to start from, appending the name to the
+ *   directory and checking that the file exists. Go up the directory hierarchy from there.
+ *
+ * @result the path to the actual package.json file that's found, or null if not.
+ */
+function _findPackagePath(realFilePathCache, name, dir) {
+
+  var h = heimdall.start('hashForDep:_findPackagePath');
+
+  try {
+    var fsRoot = pathRoot(dir);
+
+    var currPath = dir;
+
+    while (currPath !== fsRoot) {
+      var endsWithNodeModules = path.basename(currPath) === 'node_modules';
+
+      var filePath = (endsWithNodeModules ? path.join(currPath, name) : path.join(currPath, 'node_modules', name));
```

Comment:
In what scenarios would the module `name` not be found at `<some-path>/node_modules/name`?

If so, could the alg not be improved by skipping all paths who's basename is not node_modules ?

Reply:
Interestingly if `endsWithNodeModules` is always `false` all the tests pass.

Reply:
Ya the algorithm as written right seems to do duplicate IO in some scenarios.

```
'/Users/spenner/src/stefanpenner/hash-for-dep/tests/fixtures/node_modules/foo/node_modules/bar/node_modules/foo/package.json'
'/Users/spenner/src/stefanpenner/hash-for-dep/tests/fixtures/node_modules/foo/node_modules/foo/package.json'
'/Users/spenner/src/stefanpenner/hash-for-dep/tests/fixtures/node_modules/foo/node_modules/foo/package.json'
'/Users/spenner/src/stefanpenner/hash-for-dep/tests/fixtures/node_modules/foo/package.json'
```

Reply:
if `endsWithNodeModules` was true, we need to decrement the `currPath` twice, other wise the next iteration of the loop we will check the same directory again.

Reply:
> Interestingly if `endsWithNodeModules` is always `false` all the tests pass.

I definitely want to look into this. I've removed the change I made in this version (including the double dirname popping) and will go through it again to make sure it all works.

Reply:
reworked.